mayastor/src/bdev/nexus/nexus_label.rs

```diff
@@ -126,340 +141,221 @@ pub enum LabelError {
+        Ok(NexusLabel {
             status: NexusLabelStatus::Neither,
             mbr: pmbr,
             primary: header,
-            partitions: entries,
+            partitions,
             secondary: backup,
-        }
-    }
-
-    fn get_primary_data(
-        &self,
-        label: &NexusLabel,
-    ) -> Result<LabelData, LabelError> {
-        let block_size = self.bdev.block_len() as u64;
-        let blocks = Aligned::get_blocks(
-            GPTHeader::PARTITION_TABLE_SIZE as u64,
-            block_size,
-        );
-        // allocate 2 extra blocks for the MBR and GPT header respectively
-        let mut buf =
-            DmaBuf::new((blocks + 2) * block_size, self.bdev.alignment())
-                .context(WriteAlloc {
-                    name: String::from("primary"),
-                })?;
-        let mut writer = Cursor::new(buf.as_mut_slice());
-
-        // Protective MBR
-        writer.seek(SeekFrom::Start(440)).unwrap();
-        serialize_into(&mut writer, &label.mbr).context(SerializeError {})?;
-
-        // Primary GPT header
-        writer
-            .seek(SeekFrom::Start(label.primary.lba_self * block_size))
-            .unwrap();
-        serialize_into(&mut writer, &label.primary)
-            .context(SerializeError {})?;
-
-        // Primary partition table
-        writer
-            .seek(SeekFrom::Start(label.primary.lba_table * block_size))
-            .unwrap();
-        for entry in &label.partitions {
-            serialize_into(&mut writer, &entry).context(SerializeError {})?;
-        }
-
-        Ok(LabelData {
-            offset: 0,
-            buf,
         })
     }
 
-    fn get_secondary_data(
-        &self,
-        label: &NexusLabel,
-    ) -> Result<LabelData, LabelError> {
-        let block_size = self.bdev.block_len() as u64;
-        let blocks = Aligned::get_blocks(
-            GPTHeader::PARTITION_TABLE_SIZE as u64,
-            block_size,
+    // Create partition table entries for the MayaMeta and
+    // MayaData partitions based on the nexus configuration.
+    fn create_maya_partitions(
+        config: &LabelConfig,
+        header: &GPTHeader,
+        block_size: u32,
+        data_blocks: u64,
+    ) -> Result<Vec<GptEntry>, LabelError> {
+        let metadata_size = Aligned::get_blocks(
+            Nexus::METADATA_PARTITION_SIZE,
+            u64::from(block_size),
         );
-        // allocate 1 extra block for the GPT header
-        let mut buf =
-            DmaBuf::new((blocks + 1) * block_size, self.bdev.alignment())
-                .context(WriteAlloc {
-                    name: String::from("secondary"),
-                })?;
-        let mut writer = Cursor::new(buf.as_mut_slice());
+        let data = header.lba_start + metadata_size;
 
-        // Secondary partition table
-        for entry in &label.partitions {
-            serialize_into(&mut writer, &entry).context(SerializeError {})?;
+        if data > header.lba_end {
```

Comment:
I imagine this block size is the block size of this particular child and not the first one in the list?
Though IIRC there's a validation somewhere that they all need to have the same block size?

Reply:
The block size is now always obtained from the child that it pertains to. You are correct in that there is a (necessary) check to ensure that all children have the same block size, and currently lots of things will break if this is not the case. However if we are ever in a position to relax that requirement, this code will do the right thing. (Well mostly - there are still places in the labelling code that requires all children to have the same block_size - but this is no longer one of them).

Reply:
That's good. In theory it could be possible to have the nexus always configured as a 4k block device, and then that would allow having children with mixed 512 and 4k block sizes without having to do ugly things like RMW.